nncf/common/insertion_point_graph.py

```diff
@@ -0,0 +1,279 @@
+"""
+ Copyright (c) 2021 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+import functools
+import operator
+from collections import defaultdict
+from copy import deepcopy
+from enum import Enum
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Set
+
+import networkx as nx
+
+from nncf.common.graph import NNCFGraph
+from nncf.common.graph import NNCFNodeExpression
+from nncf.common.graph import NNCFNodeName
+from nncf.common.graph import NodeExpression
+from nncf.common.hardware.config import HWConfig
+
+
+class InsertionPointGraphNodeType(Enum):
+    PRE_HOOK = 0
+    POST_HOOK = 1
+    OPERATOR = 2
+
+
+class PreHookInsertionPoint:
+    def __init__(self, target_node_name: str, input_port_id: int):
+        self.target_node_name = target_node_name
+        self.input_port_id = input_port_id
+
+
+class PostHookInsertionPoint:
+    def __init__(self, target_node_name: str):
+        self.target_node_name = target_node_name
+
+
+class InsertionPointGraph(nx.DiGraph):
+    """
+    This graph is built from the NNCFGraph representation of the model control flow graph and adds ephemeral
+    "insertion point nodes" into the NNCF model graph representation corresponding to operator pre- and
+    post-hooks. Module pre-op and post-op insertion points are currently not reflected here, but they are
+    probably not required for quantizing activations, for which the quantizer propagation makes sense.
+    This "insertion point graph" representation is useful for quantizer propagation and for referencing
+    the compression algorithm hooks to the model operations to which they are applied to.
+    """
+    NODE_TYPE_NODE_ATTR = "node_type"
```

Comment:
Could you please use `'` instead of `"` for consistency here and below?

Reply:
Done